File: frontend/src/models/MemorySection.js
/**
 * Domain model representing a memory section.
 * Mirrors backend MemorySection model exactly.
 * 
 * Strict OOP principles:
 * - Immutable with Object.freeze()
 * - Deep frozen content
 * - Validation in constructor
 */
export class MemorySection {
  /**
   * Create a new MemorySection
   * @param {Object} params Section parameters
   * @param {string} params.sectionName Section name (e.g., 'findings', 'context_chain')
   * @param {*} params.content Section content (any serializable data)
   * @param {string|Date} params.timestamp Snapshot timestamp
   * @param {Object} params.metadata Additional metadata
   */
  constructor({ sectionName, content, timestamp, metadata = {} }) {
    // Validation
    if (!sectionName || typeof sectionName !== 'string') {
      throw new Error(`MemorySection: sectionName is required and must be a string, got ${typeof sectionName}`);
    }
    if (content === undefined || content === null) {
      throw new Error('MemorySection: content is required');
    }
    if (!timestamp) {
      throw new Error('MemorySection: timestamp is required');
    }
    if (metadata && typeof metadata !== 'object') {
      throw new Error(`MemorySection: metadata must be an object, got ${typeof metadata}`);
    }

    this._sectionName = sectionName;
    this._content = this._deepFreeze(content);
    this._timestamp = this._parseTimestamp(timestamp);
    this._metadata = Object.freeze({ ...metadata });

    Object.freeze(this);
  }

  // Getters
  get sectionName() { return this._sectionName; }
  get content() { return this._content; }
  get timestamp() { return this._timestamp; }
  get metadata() { return this._metadata; }

  /**
   * Check if section is empty
   * @returns {boolean}
   */
  isEmpty() {
    if (this._content === null || this._content === undefined) {
      return true;
    }
    if (Array.isArray(this._content)) {
      return this._content.length === 0;
    }
    if (typeof this._content === 'object') {
      return Object.keys(this._content).length === 0;
    }
    return false;
  }

  /**
   * Get content size
   * @returns {number|null}
   */
  getSize() {
    if (Array.isArray(this._content)) {
      return this._content.length;
    }
    if (typeof this._content === 'object' && this._content !== null) {
      return Object.keys(this._content).length;
    }
    return null;
  }

  /**
   * Get formatted timestamp
   * @returns {string}
   */
  getFormattedTimestamp() {
    return this._timestamp.toLocaleString();
  }

  /**
   * Serialize to JSON
   * @returns {Object}
   */
  toJSON() {
    return {
      section_name: this._sectionName,
      content: this._content,
      timestamp: this._timestamp.toISOString(),
      metadata: { ...this._metadata },
      size: this.getSize(),
      empty: this.isEmpty()
    };
  }

  /**
   * Create MemorySection from JSON
   * @param {Object} json JSON data
   * @returns {MemorySection}
   */
  static fromJSON(json) {
    if (!json || typeof json !== 'object') {
      throw new Error('MemorySection.fromJSON: json must be an object');
    }

    return new MemorySection({
      sectionName: json.section_name || json.sectionName,
      content: json.content,
      timestamp: json.timestamp,
      metadata: json.metadata || {}
    });
  }

  /**
   * Parse timestamp from various formats
   * @private
   */
  _parseTimestamp(value) {
    if (value instanceof Date) {
      return value;
    }
    if (typeof value === 'number') {
      return new Date(value);
    }
    if (typeof value === 'string') {
      const parsed = new Date(value);
      if (isNaN(parsed.getTime())) {
        throw new Error(`MemorySection: Invalid timestamp '${value}'`);
      }
      return parsed;
    }
    throw new Error(`MemorySection: Invalid timestamp type ${typeof value}`);
  }

  /**
   * Deep freeze nested structures
   * @private
   */
  _deepFreeze(obj) {
    if (obj === null || obj === undefined) {
      return obj;
    }

    if (Array.isArray(obj)) {
      const frozen = obj.map(item => this._deepFreeze(item));
      return Object.freeze(frozen);
    }

    if (typeof obj === 'object') {
      const frozen = {};
      for (const key in obj) {
        if (obj.hasOwnProperty(key)) {
          frozen[key] = this._deepFreeze(obj[key]);
        }
      }
      return Object.freeze(frozen);
    }

    return obj;
  }
}

File: app/models/memory_section.rb
# frozen_string_literal: true

# Domain model representing a memory section snapshot.
# Captures state of a specific memory section at a point in time.
#
# Strict OOP principles:
# - Immutable value object
# - Validation in constructor
# - No hash access patterns
class MemorySection
  attr_reader :section_name, :content, :timestamp, :metadata

  # Initialize a new memory section
  # @param section_name [String, Symbol] Section name (e.g., :findings, :context_chain)
  # @param content [Object] Section content (array, hash, or other serializable data)
  # @param timestamp [Time, String] Snapshot timestamp
  # @param metadata [Hash] Additional metadata
  def initialize(section_name:, content:, timestamp:, metadata: {})
    raise ArgumentError, "section_name is required" if section_name.nil? || section_name.to_s.empty?
    raise ArgumentError, "content is required" if content.nil?
    raise ArgumentError, "timestamp is required" if timestamp.nil?
    raise ArgumentError, "metadata must be a Hash" unless metadata.nil? || metadata.is_a?(Hash)

    @section_name = section_name.to_sym
    @content = deep_freeze(content)
    @timestamp = parse_time(timestamp)
    @metadata = (metadata || {}).dup.freeze
    
    freeze
  end

  # Check if section is empty
  # @return [Boolean]
  def empty?
    @content.nil? || 
    (@content.respond_to?(:empty?) && @content.empty?) ||
    (@content.respond_to?(:size) && @content.size == 0)
  end

  # Get content size
  # @return [Integer, nil]
  def size
    return nil unless @content.respond_to?(:size)
    @content.size
  end

  # Serialize to hash for API responses
  # @return [Hash]
  def to_h
    {
      section_name: @section_name.to_s,
      content: @content,
      timestamp: @timestamp.iso8601,
      metadata: @metadata,
      size: size,
      empty: empty?
    }
  end
  alias_method :to_json, :to_h

  # Create memory section from hash
  # @param data [Hash] Memory section data
  # @return [MemorySection]
  def self.from_h(data)
    raise ArgumentError, "data must be a Hash" unless data.is_a?(Hash)
    
    new(
      section_name: data[:section_name] || data["section_name"],
      content: data[:content] || data["content"],
      timestamp: data[:timestamp] || data["timestamp"],
      metadata: data[:metadata] || data["metadata"] || {}
    )
  end

  private

  # Parse time from various formats
  # @param value [Time, String, Integer] Time value
  # @return [Time]
  def parse_time(value)
    return value if value.is_a?(Time)
    return Time.at(value) if value.is_a?(Integer)
    return Time.parse(value) if value.is_a?(String)
    raise ArgumentError, "Invalid time value: #{value}"
  end

  # Deep freeze nested structures
  # @param obj [Object] Object to freeze
  # @return [Object] Frozen object
  def deep_freeze(obj)
    case obj
    when Hash
      obj.each { |k, v| deep_freeze(v) }
      obj.freeze
    when Array
      obj.each { |v| deep_freeze(v) }
      obj.freeze
    else
      obj.freeze if obj.respond_to?(:freeze)
      obj
    end
  end
end








